HW2/P3/mandelbrot.pyx

```diff
@@ -5,73 +5,168 @@ import numpy
 cimport AVX
 from cython.parallel import prange
 
-
 cdef np.float64_t magnitude_squared(np.complex64_t z) nogil:
-    return z.real * z.real + z.imag * z.imag
+	return z.real * z.real + z.imag * z.imag
+
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
+
+
+cpdef mandelbrot_AVX(np.complex64_t [:, :] in_coords,
+				np.uint32_t [:, :] out_counts,
+				int n_threads,
+				int max_iterations=511):
+	cdef:
+		int i, j, iter, k
+		np.float32_t [:, :] in_coords_real, in_coords_imag
+		AVX.float8 c_real, c_imag, iter_AVX, z_mod
+		AVX.float8 mask, z_real, z_imag, z_real_temp
+
+
+	in_coords_real = np.real(in_coords)
+	in_coords_imag = np.imag(in_coords)
+
+
+	# parallize the rows with prange
+	for i in prange(in_coords.shape[0], nogil = True, schedule = 'static', chunksize =1,
+	 				num_threads = n_threads):
+
+		# we now go by steps of 8 within the loop because each AVX takes care of 8 values
+		for j in range(0,in_coords.shape[1],8):
+			# define c_real (in the AVX 7 to 0 format)
+			c_real = AVX.make_float8(in_coords_real[i, j+7], in_coords_real[i, j+6], in_coords_real[i, j+5],
+			in_coords_real[i, j+4], in_coords_real[i, j+3], in_coords_real[i, j+2], in_coords_real[i, j+1],
+			in_coords_real[i, j+0])
+			# define c_imag (in the AVX 7 to 0 format)
+			c_imag = AVX.make_float8(in_coords_imag[i, j+7], in_coords_imag[i, j+6], in_coords_imag[i, j+5],
+			in_coords_imag[i, j+4], in_coords_imag[i, j+3], in_coords_imag[i, j+2], in_coords_imag[i, j+1],
+			in_coords_imag[i, j+0])
+			# initialize z_real, z_imag and iter_AVX which is the output of the image
+			z_real = AVX.float_to_float8(0.)
+			z_imag = AVX.float_to_float8(0.)
+			iter_AVX = AVX.float_to_float8(0.)
+
+
+			# now iterate and count teh number of iterations
+			for iter in range(max_iterations):
+
+				# z = a+ib, real z is a^2-(b^2-cr)
+				# temporary because still needed to compute z_imag
+				z_real_temp = AVX.sub(AVX.fmadd(z_real, z_real, c_real),
+													 AVX.mul(z_imag, z_imag))
+				# imag 2ab + ci
+				z_imag = AVX.fmadd(AVX.mul(AVX.float_to_float8(2.),z_real), z_imag, c_imag)
+				z_real = z_real_temp
+
+				# find the squared magnitude of z
+				# |z|^2 = x^2 + y^2 where z = x + iy
+				z_mod = AVX.add(AVX.mul(z_real, z_real), AVX.mul(z_imag, z_imag))
+
+				# mask repeat while z_mod < 4
+				mask = AVX.less_than(z_mod, AVX.float_to_float8(4.))
+
+				#avoid unecessary iteration
+				#when all the mask is 0 it means that there is no need to update
+				if AVX.signs(mask) == 0:
+					break
+
+				# update iter_AVX:
+				# AVX.bitwise_and(mask,  AVX.float_to_float8(1.)) gives a one only where mask is not zero
+				iter_AVX = AVX.add(iter_AVX, AVX.bitwise_and(mask,  AVX.float_to_float8(1.)))
+
+
+			# Assign the iterations to out_counts
+			for k in range(8):
+				out_counts[i,j+k]= <np.uint32_t> (<np.float32_t *> &iter_AVX)[k]
+
```

Comment:
Does it work? When I used the below, i encountered the error (I didn't remember what error it was.)
for k in range(8):
+				out_counts[i,j+k]= <np.uint32_t> (<np.float32_t *> &iter_AVX)[k]
So, in my case, I used the function that is recommended to be used on the piazza.